// @dapr/dapr/implementation/Server/DaprServer.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const CommunicationProtocol_enum_1 = __importDefault(require("../../enum/CommunicationProtocol.enum"));
const GRPCServer_1 = __importDefault(require("./GRPCServer/GRPCServer"));
const pubsub_1 = __importDefault(require("./GRPCServer/pubsub"));
const binding_1 = __importDefault(require("./GRPCServer/binding"));
const invoker_1 = __importDefault(require("./GRPCServer/invoker"));
const actor_1 = __importDefault(require("./GRPCServer/actor"));
const HTTPServer_1 = __importDefault(require("./HTTPServer/HTTPServer"));
const pubsub_2 = __importDefault(require("./HTTPServer/pubsub"));
const binding_2 = __importDefault(require("./HTTPServer/binding"));
const invoker_2 = __importDefault(require("./HTTPServer/invoker"));
const actor_2 = __importDefault(require("./HTTPServer/actor"));
const __1 = require("../..");
const Settings_util_1 = require("../../utils/Settings.util");
const Logger_1 = require("../../logger/Logger");
class DaprServer {
    constructor(serverHost, serverPort, daprHost, daprPort, communicationProtocol = CommunicationProtocol_enum_1.default.HTTP, clientOptions = {}) {
        this.serverHost = serverHost !== null && serverHost !== void 0 ? serverHost : Settings_util_1.Settings.getDefaultHost();
        this.serverPort = serverPort !== null && serverPort !== void 0 ? serverPort : Settings_util_1.Settings.getDefaultAppPort(communicationProtocol);
        this.daprHost = daprHost !== null && daprHost !== void 0 ? daprHost : Settings_util_1.Settings.getDefaultHost();
        this.daprPort = daprPort !== null && daprPort !== void 0 ? daprPort : Settings_util_1.Settings.getDefaultPort(communicationProtocol);
        this.logger = new Logger_1.Logger("DaprServer", "DaprServer", clientOptions.logger);
        // Create a client to interface with the sidecar from the server side
        this.client = new __1.DaprClient(daprHost, daprPort, communicationProtocol, clientOptions);
        // If DAPR_SERVER_PORT was not set, we set it
        process.env.DAPR_SERVER_PORT = this.serverPort;
        process.env.DAPR_CLIENT_PORT = this.daprPort;
        // Validation on port
        if (!/^[0-9]+$/.test(this.serverPort)) {
            throw new Error("DAPR_INCORRECT_SERVER_PORT");
        }
        if (!/^[0-9]+$/.test(this.daprPort)) {
            throw new Error("DAPR_INCORRECT_SIDECAR_PORT");
        }
        // Builder
        switch (communicationProtocol) {
            case CommunicationProtocol_enum_1.default.GRPC: {
                const server = new GRPCServer_1.default(this.client);
                this.daprServer = server;
                this.pubsub = new pubsub_1.default(server);
                this.binding = new binding_1.default(server);
                this.invoker = new invoker_1.default(server);
                this.actor = new actor_1.default(server);
                break;
            }
            case CommunicationProtocol_enum_1.default.HTTP:
            default: {
                const server = new HTTPServer_1.default(this.client);
                this.daprServer = server;
                this.pubsub = new pubsub_2.default(server);
                this.binding = new binding_2.default(server);
                this.invoker = new invoker_2.default(server);
                this.actor = new actor_2.default(server, this.client);
                break;
            }
        }
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            // First start the server as we need to initialize routes for PubSub, Bindings, ...
            yield this.daprServer.start(this.serverHost, this.serverPort.toString());
            // Ensure our sidecar starts and the client is ready
            yield this.client.start();
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.daprServer.stop();
        });
    }
    getDaprClient() {
        return this.daprServer;
    }
    getDaprHost() {
        return this.daprHost;
    }
    getDaprPort() {
        return this.daprPort;
    }
}
exports.default = DaprServer;
