// @dapr/dapr/implementation/Client/HTTPClient/state.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// https://docs.dapr.io/reference/api/state_api/
class HTTPClientState {
    constructor(client) {
        this.client = client;
    }
    save(storeName, stateObjects) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/state/${storeName}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(stateObjects),
            });
        });
    }
    get(storeName, key) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.execute(`/state/${storeName}/${key}`);
            return result;
        });
    }
    getBulk(storeName, keys, parallelism = 10, metadata = "") {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.execute(`/state/${storeName}/bulk${metadata ? `?${metadata}` : ""}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    keys,
                    parallelism, // the number of parallel operations executed on the state store for a get operation
                }),
            });
            return result;
        });
    }
    delete(storeName, key) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/state/${storeName}/${key}`, {
                method: "DELETE",
            });
        });
    }
    transaction(storeName, operations = [], metadata = null) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/state/${storeName}/transaction`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    operations,
                    metadata,
                }),
            });
        });
    }
    query(storeName, query) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.executeWithApiVersion("v1.0-alpha1", `/state/${storeName}/query`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(Object.assign({}, query)),
            });
            return result;
        });
    }
}
exports.default = HTTPClientState;
