// skywalking-backend-js/lib/trace/context/Segment.js
"use strict";
/*!
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ID_1 = tslib_1.__importDefault(require("../../trace/ID"));
var NewID_1 = tslib_1.__importDefault(require("../../trace/NewID"));
var Segment = /** @class */ (function () {
    function Segment() {
        this.segmentId = new ID_1.default();
        this.spans = [];
        this.relatedTraces = [new NewID_1.default()];
        this.references = [];
    }
    Segment.prototype.archive = function (span) {
        this.spans.push(span);
    };
    Segment.prototype.relate = function (id) {
        if (this.relatedTraces[0] instanceof NewID_1.default) {
            this.relatedTraces.shift();
        }
        if (!this.relatedTraces.includes(id)) {
            this.relatedTraces.push(id);
        }
    };
    Segment.prototype.refer = function (ref) {
        if (!this.references.includes(ref)) {
            this.references.push(ref);
        }
        return this;
    };
    return Segment;
}());
exports.default = Segment;
