// @dapr/dapr/implementation/Server/HTTPServer/HTTPServerImpl.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const SubscribedMessageHttpResponse_enum_1 = __importDefault(require("../../../enum/SubscribedMessageHttpResponse.enum"));
const Logger_1 = require("../../../logger/Logger");
class HTTPServerImpl {
    constructor(server, loggerOptions) {
        this.PUBSUB_DEFAULT_ROUTE_NAME = "default";
        this.PUBSUB_DEFAULT_ROUTE_NAME_DEADLETTER = "deadletter";
        this.server = server;
        this.logger = new Logger_1.Logger("HTTPServer", "HTTPServerImpl", loggerOptions);
        this.pubSubSubscriptions = {};
    }
    /**
     * When we subscribe, we subscribe to a topic
     * For this topic we can define "routes" which route to a certain callback depending on the event content
     * Each of these topics are handled by a EventHandler but there can be multiple handlers per pubsubname-topic-route combination
     *
     * We don't create the EventHandlers here but we ensure that the routes are registered and can receive POST events
     * -> we create POST /<route> endpoints for each, but we create them uniquely!
     * -> to ensure uniqueness, we just check if this.pubsubRouteEventHandlers[route] is set
     *
     * @param pubSubName
     * @param topicName
     * @param cb
     * @param options
     */
    registerPubsubSubscription(pubsubName, topic, options = {}) {
        if (this.pubSubSubscriptions[pubsubName] &&
            this.pubSubSubscriptions[pubsubName][topic] &&
            this.pubSubSubscriptions[pubsubName][topic]) {
            throw new Error(`The topic '${topic}' is already being subscribed to on PubSub '${pubsubName}', there can only be one topic registered.`);
        }
        // Create pubsub subscription it if it doesn't exist
        if (!this.pubSubSubscriptions[pubsubName]) {
            this.pubSubSubscriptions[pubsubName] = {};
        }
        // Add the routes and dapr representation to the topic
        if (!this.pubSubSubscriptions[pubsubName][topic]) {
            this.pubSubSubscriptions[pubsubName][topic] = {
                routes: this.generatePubSubSubscriptionTopicRoutes(pubsubName, topic, options),
                dapr: this.generateDaprPubSubSubscription(pubsubName, topic, options),
            };
        }
        // Listen to the created endpoints on the HTTP Server
        for (const route of Object.keys(this.pubSubSubscriptions[pubsubName][topic].routes)) {
            const routeObj = this.pubSubSubscriptions[pubsubName][topic].routes[route];
            // Add a server POST handler
            this.server.post(`/${routeObj.path}`, (req, res) => __awaiter(this, void 0, void 0, function* () {
                var _a;
                // @ts-ignore
                // Parse the data of the body, we prioritize fetching the data key in body if possible
                // i.e. Redis returns { data: {} } and other services return {}
                // @todo: This will be deprecated in an upcoming major version and only req.body will be returned
                const data = ((_a = req === null || req === void 0 ? void 0 : req.body) === null || _a === void 0 ? void 0 : _a.data) || (req === null || req === void 0 ? void 0 : req.body);
                const headers = req.headers;
                // Process our callback
                try {
                    const eventHandlers = routeObj.eventHandlers;
                    yield Promise.all(eventHandlers.map((cb) => cb(data, headers)));
                }
                catch (e) {
                    this.logger.error(`[route-${routeObj.path}] Message processing failed, dropping: ${e}`);
                    return res.send({ status: SubscribedMessageHttpResponse_enum_1.default.DROP });
                }
                // Let Dapr know that the message was processed correctly
                this.logger.debug(`[route-${routeObj.path}] Ack'ing the message`);
                return res.send({ status: SubscribedMessageHttpResponse_enum_1.default.SUCCESS });
            }));
        }
        this.logger.info(`[Topic = ${topic}] Registered Subscription with routes: ${Object.keys(this.pubSubSubscriptions[pubsubName][topic].routes).join(", ")}`);
    }
    registerPubSubSubscriptionEventHandler(pubsubName, topic, route, cb) {
        route = this.generatePubSubSubscriptionTopicRouteName(route);
        this.pubSubSubscriptions[pubsubName][topic].routes[route !== null && route !== void 0 ? route : this.PUBSUB_DEFAULT_ROUTE_NAME].eventHandlers.push(cb);
    }
    generatePubSubSubscriptionTopicRouteName(route = "default") {
        return (route || this.PUBSUB_DEFAULT_ROUTE_NAME).replace("/", "");
    }
    generatePubSubSubscriptionTopicRoutes(pubsubName, topic, options = {}) {
        var _a;
        const routes = {};
        // options.route == DaprPubSubRouteType
        if (typeof options.route === "object") {
            // Add default
            if (options.route.default) {
                const routeName = this.generatePubSubSubscriptionTopicRouteName(options.route.default);
                routes[routeName] = {
                    eventHandlers: [],
                    path: this.generatePubsubPath(pubsubName, topic, routeName),
                };
            }
            // Add rules
            if (options.route.rules) {
                for (const rule of options.route.rules) {
                    if (!routes[rule.path]) {
                        const routeName = this.generatePubSubSubscriptionTopicRouteName(rule.path);
                        routes[routeName] = {
                            eventHandlers: [],
                            path: this.generatePubsubPath(pubsubName, topic, routeName),
                        };
                    }
                }
            }
        }
        // options.route == String | undefined
        else {
            const routeName = this.generatePubSubSubscriptionTopicRouteName(options === null || options === void 0 ? void 0 : options.route);
            routes[routeName] = {
                eventHandlers: [],
                path: this.generatePubsubPath(pubsubName, topic, routeName),
            };
        }
        // Deadletter Support
        if (options.deadLetterTopic || options.deadLetterCallback) {
            const routeName = this.generatePubSubSubscriptionTopicRouteName((_a = options === null || options === void 0 ? void 0 : options.deadLetterTopic) !== null && _a !== void 0 ? _a : this.PUBSUB_DEFAULT_ROUTE_NAME_DEADLETTER);
            // Initialize the route
            routes[routeName] = {
                eventHandlers: [],
                path: this.generatePubsubPath(pubsubName, topic, routeName),
            };
            // Add a callback if we have one provided
            if (options.deadLetterCallback) {
                routes[routeName].eventHandlers.push(options.deadLetterCallback);
            }
        }
        return routes;
    }
    generateDaprSubscriptionRoute(pubsubName, topic, route = this.PUBSUB_DEFAULT_ROUTE_NAME) {
        return `/${this.generatePubsubPath(pubsubName, topic, route)}`;
    }
    /**
     * Generate a subscription object that will be used in the /dapr/subscribe endpoint
     * this will let dapr know that we have subscriptions and how they map to routes / deadletter
     *
     * Important: we internally translate the provided /example to -> /<pubsubname>-<topic>-example
     *            or if empty to /<pubsubname>-<topic>-default
     *            this is to ensure that HTTP Server endpoints are unique
     *
     * @param pubsubName
     * @param topic
     * @param options
     * @returns
     */
    generateDaprPubSubSubscription(pubsubName, topic, options = {}) {
        var _a, _b, _c;
        // Process metadata
        let metadata;
        if (options.metadata) {
            metadata = {};
            for (const [key, value] of Object.entries(options.metadata)) {
                metadata[key] = JSON.stringify(value);
            }
        }
        // Process the route
        if (!options || !(options === null || options === void 0 ? void 0 : options.route)) {
            return {
                pubsubname: pubsubName,
                topic: topic,
                metadata: metadata,
                route: this.generateDaprSubscriptionRoute(pubsubName, topic),
                deadLetterTopic: options.deadLetterTopic,
            };
        }
        else if (typeof options.route === "string") {
            return {
                pubsubname: pubsubName,
                topic: topic,
                metadata: metadata,
                route: this.generateDaprSubscriptionRoute(pubsubName, topic, options.route),
                deadLetterTopic: options.deadLetterTopic,
            };
        }
        else {
            return {
                pubsubname: pubsubName,
                topic: topic,
                metadata: metadata,
                routes: options.route && {
                    default: this.generateDaprSubscriptionRoute(pubsubName, topic, (_a = options.route) === null || _a === void 0 ? void 0 : _a.default),
                    rules: (_c = (_b = options.route) === null || _b === void 0 ? void 0 : _b.rules) === null || _c === void 0 ? void 0 : _c.map((rule) => ({
                        match: rule.match,
                        path: this.generateDaprSubscriptionRoute(pubsubName, topic, rule.path),
                    })),
                },
                deadLetterTopic: options.deadLetterTopic,
            };
        }
    }
    generateDaprPubSubSubscriptionList() {
        const dapr = [];
        for (const pubsub of Object.keys(this.pubSubSubscriptions)) {
            for (const topic of Object.keys(this.pubSubSubscriptions[pubsub])) {
                dapr.push(this.pubSubSubscriptions[pubsub][topic].dapr);
            }
        }
        return dapr;
    }
    /**
     * We generate a event handler key based on the path or the route
     * If the route is just a string, that is the path
     * Else the path is configured through a rule of DaprPubSubRuleType
     *
     * @param pubsubName
     * @param topic
     * @param route
     * @returns
     */
    generatePubsubPath(pubsubName, topic, route) {
        let routeParsed = "";
        // First parse the route based on if it was a Rule or a String
        if (!route) {
            routeParsed = this.PUBSUB_DEFAULT_ROUTE_NAME;
        }
        else {
            routeParsed = route;
        }
        // Then, process it
        // Remove leading slashes
        if (routeParsed.startsWith("/")) {
            routeParsed = routeParsed.replace("/", ""); // will only remove first occurence
        }
        return `${pubsubName.toLowerCase()}--${topic.toLowerCase()}--${routeParsed}`;
    }
}
exports.default = HTTPServerImpl;
