// @dapr/dapr/actors/client/ActorClient/ActorClientHTTP.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// https://docs.dapr.io/reference/api/actors_api/
class ActorClientHTTP {
    constructor(client) {
        this.client = client;
    }
    invoke(actorType, actorId, methodName, body) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/method/${methodName}`, {
                method: "POST",
                body,
            });
            return result;
        });
    }
    stateTransaction(actorType, actorId, operations) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/state`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(operations),
            });
        });
    }
    stateGet(actorType, actorId, key) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/state/${key}`);
            return result;
        });
    }
    registerActorReminder(actorType, actorId, name, reminder) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/reminders/${name}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    period: reminder.period.toString().toLocaleLowerCase().replace("pt", ""),
                    dueTime: (_b = (_a = reminder === null || reminder === void 0 ? void 0 : reminder.dueTime) === null || _a === void 0 ? void 0 : _a.toString()) === null || _b === void 0 ? void 0 : _b.toLocaleLowerCase().replace("pt", ""),
                    ttl: (_d = (_c = reminder === null || reminder === void 0 ? void 0 : reminder.ttl) === null || _c === void 0 ? void 0 : _c.toString()) === null || _d === void 0 ? void 0 : _d.toLocaleLowerCase().replace("pt", ""),
                    data: reminder.data,
                }),
            });
        });
    }
    reminderGet(actorType, actorId, name) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/reminders/${name}`);
            return result;
        });
    }
    unregisterActorReminder(actorType, actorId, name) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/reminders/${name}`, {
                method: "DELETE",
            });
        });
    }
    registerActorTimer(actorType, actorId, name, timer) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/timers/${name}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    period: timer.period.toString().toLocaleLowerCase().replace("pt", ""),
                    dueTime: (_b = (_a = timer === null || timer === void 0 ? void 0 : timer.dueTime) === null || _a === void 0 ? void 0 : _a.toString()) === null || _b === void 0 ? void 0 : _b.toLocaleLowerCase().replace("pt", ""),
                    ttl: (_d = (_c = timer === null || timer === void 0 ? void 0 : timer.ttl) === null || _c === void 0 ? void 0 : _c.toString()) === null || _d === void 0 ? void 0 : _d.toLocaleLowerCase().replace("pt", ""),
                    data: timer.data,
                    callback: timer.callback,
                }),
            });
        });
    }
    unregisterActorTimer(actorType, actorId, name) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/actors/${actorType}/${actorId.getId()}/timers/${name}`, {
                method: "DELETE",
            });
        });
    }
    deactivate(actorType, actorId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.execute(`/actors/${actorType}/${actorId}`, {
                method: "DELETE",
            });
        });
    }
    getActors() {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.execute(`replace('/v1.0', '')}/dapr/config`);
            return result;
        });
    }
}
exports.default = ActorClientHTTP;
